Dynatrace/dynatrace-operator
# Add KSPM DaemonSet reconciler

## Description
[DAQ-1448](https://dt-rnd.atlassian.net/browse/DAQ-1448)

Adds logic to deploy the node-collector `DaemonSet` for the KSPM feature.
(also includes some small fixes)
 

## How can this be tested?

- Image to use: `us-central1-docker.pkg.dev/cloud-platform-207208/chmu/k8s-node-config-collector`
   - The current default is not yet available 
- Use the `spec.activeGate.tlsSecretName`
   - I tested it, reason for [Fix cert handling](https://github.com/Dynatrace/dynatrace-operator/pull/3997/commits/e6f1ddca7ee12efa1e5cb0d23fe91f7949eb7b07), but double checking would be nice :D
- Use the `spec.activeGate.customProperties`
   - you have to enable the feature

Example `DynaKube`:
```
apiVersion: dynatrace.com/v1beta3
kind: DynaKube
metadata:
  name: dk
  namespace: dynatrace
spec:
  apiUrl: https://<tenant>.dev.dynatracelabs.com/api
  kspm:
    enabled: true
  activeGate:
    tlsSecretName: my-secret
    capabilities:
      - kubernetes-monitoring
    customProperties:
      value: |
        [kubernetes_monitoring]
        kubernetes_configuration_dataset_pipeline_enabled = true
        kubernetes_configuration_dataset_pipeline_include_node_config = true
  templates:
     kspmNodeConfigurationCollector:
       imageRef:
         repository: us-central1-docker.pkg.dev/cloud-platform-207208/chmu/k8s-node-config-collector
         tag: latest
```

Things to check:
- `DaemonSet` gets created
   - Related pods come up
   - Related pod respect what is configured in `templates.kspmNodeConfigurationCollector`
- `DaemonSet` gets recreated if the token-secret changes
- `Conditions` on the `DynaKube` are present and updated accordingly

[DAQ-1448]: https://dt-rnd.atlassian.net/browse/DAQ-1448?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
